wip/3mdeb/qubes-wrapper

Asiderr opened the pull request:
This patch is adding the fwupd wrapper for Qubes OS. The wrapper provides fwupd functionalities for Qubes R4.1.
It creates three packages (two RPMs and one deb):
fwupd-qubes-dom0 (RPM)
fwupd-qubes-vm (RPM)
fwupd-qubes-vm-whonix (deb)
More information about the wrapper could be found in the [contrib/qubes/README.md](https://github.com/3mdeb/fwupd/tree/wip/3mdeb/qubes-wrapper/contrib/qubes)


Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [x] Documentation

hughsie commented on contrib/fwupd.spec.in:
```diff
@@ -181,6 +183,30 @@ This provides the optional package which is only required on hardware that
+Requires:   cabextract
+Requires:   fwupd
+Requires:   gpg
+%if 0%{?fedora} <= 25
+Requires:   python36
+%endif
```
I don't think we care about obsolete versions

Asiderr replied:
Qubes OS R4.0 Dom0 is based on Fedora 25. If we will drop support for Qubes R4.0, Fedora 25 related lines should be deleted.

Asiderr replied:
Deleted. I've left support only for Qubes R4.1

hughsie commented on contrib/fwupd.spec.in:
```diff
@@ -484,6 +510,52 @@ done
+
+%post qubes-dom0
+# Replace the python shebangs if fedora is older then 25
+%if 0%{?fedora} <= 25
+sed -i 's/python3/python36/g' %FWUPD_QUBES_DIR/src/qubes_fwupdmgr.py
+sed -i 's/python3/python36/g' %FWUPD_QUBES_DIR/src/fwupd_receive_updates.py
+sed -i 's/python3/python36/g' %FWUPD_QUBES_DIR/src/qubes_fwupd_heads.py
+sed -i 's/python3/python36/g' %FWUPD_QUBES_DIR/test/test_qubes_fwupdmgr.py
+sed -i 's/python3/python36/g' %FWUPD_QUBES_DIR/test/test_qubes_fwupd_heads.py
+sed -i 's/python3/python36/g' %FWUPD_QUBES_DIR/test/fwupd_logs.py
+%endif
```
same; Fedora 25 is EOL and dead dead dead, no?

Asiderr replied:
As above.

Asiderr replied:
Deleted. I've left support only for Qubes R4.1

hughsie commented on contrib/fwupd.spec.in:
```diff
@@ -484,6 +510,52 @@ done
+    qvm-run --pass-io sys-firewall 'gpg --import %PKI/GPG-KEY-Linux-Vendor-Firmware-Service'
+fi
+if xl list sys-whonix &>/dev/null; then
+    qvm-run --pass-io sys-whonix 'rm /home/user/QubesIncoming/dom0/GPG-KEY-Linux-Vendor-Firmware-Service'
+    qvm-copy-to-vm sys-whonix %PKI/GPG-KEY-Linux-Vendor-Firmware-Service
+    qvm-run --pass-io sys-whonix 'gpg --import /home/user/QubesIncoming/dom0/GPG-KEY-Linux-Vendor-Firmware-Service'
```
Just the GPG keys? I think you want the PKCS ones too, no?

Asiderr replied:
GPG keys are used three times to verify the metadata and firmware update files. The first time, they are used when the files are downloaded, the second time after copying the files from UpdateVM to Dom0, and the third time when Dom0 copies files to sys-usb VM. For now, we are not using the PKCS keys, but if we want to use `jcat-tool` they are available in standard fwupd directory `/etc/pki/fwupd`

Asiderr replied:
I deleted these lines. GPG verification has been replaced with jcat verification.

hughsie commented on contrib/qubes/src/fwupd-dom0-update:
```diff
@@ -0,0 +1,190 @@
+escape_args() {
+    local eargs=""
+
+    for arg in "$@"; do
+    printf -v eargs '%s%q ' "$eargs" "$arg"
+    done
```
does this file have to be shell? I'd be much more comfortable if it was python...

Asiderr replied:
I rewrote the shell scripts to the Python.

hughsie commented on contrib/qubes/src/fwupd_receive_updates.py:
```diff
@@ -0,0 +1,346 @@
+        Keyword argument:
+        file_path -- absolute path to inspected file
+        """
+        cmd_gpg = [
+            "gpg",
+            "--verify",
+            f"{file_path}.asc",
+            f"{file_path}",
+        ]
```
I think hardcoding GPG is probably dangerous; I think going forward we'd probably drop the standalone keys and just use the jcat file.

Asiderr replied:
It could be changed to `jcat-tool` if we will drop the Qubes R4.0 support. I'm not sure if there is `libjcat` package for Fedora 25.

Asiderr replied:
I've added the [`_jcat_verification`](https://github.com/fwupd/fwupd/pull/2710/files#diff-75406b905f82fd50a6d9b1cfe76b42b735a197d178d2cf9c4f1fc95fb2c16981R185) method, which is used for metadata verification.

hughsie commented on contrib/qubes/test/logs/get_updates.log:
```diff
@@ -0,0 +1,15 @@
+   SHA1 checksum:	 490be5c0b13ca4a3f169bf8bc682ba127b8f7b96
+   Description: This release fixes prevents the firmware returning an error when the remote SHA1 hash was never sent.
+	
+======================================================
```
quite a few of these files are missing the newline char at the end...

Asiderr replied:
Fixed.

hughsie commented on contrib/qubes/src/updatevm/fwupd-download-updates.sh:
```diff
@@ -0,0 +1,122 @@
+    echo "Downloading metadata."
+    rm -rf $FWUPD_UPDATEVM_DIR/metadata/*
+    wget -P $FWUPD_UPDATEVM_DIR/metadata $URL
+    wget -P $FWUPD_UPDATEVM_DIR/metadata $URL.jcat
+    wget -P $FWUPD_UPDATEVM_DIR/metadata $URL.asc
```
both?

Asiderr replied:
We need both files to cover fwupd versions from 0.9.5 to the current one.

superm1 replied:
Why support such old fwupd versions?

Asiderr replied:
Ferdora 25 is base for dom0 in Qubes R4.0 (https://www.qubes-os.org/doc/supported-versions/) and it provides fwupd 0.9.5.

Asiderr replied:
The file doesn't exist. It's replaced with the python script. There is no need to download the GPG signature, thus I left only `jcat` file. The wrapper works correctly with the current libjcat master and the changes connected to the https://github.com/hughsie/libjcat/issues/48 and https://github.com/hughsie/libjcat/issues/49 need to be available on Fedora 32.

hughsie replied:
libjcat update just went stable for f32: https://bodhi.fedoraproject.org/updates/FEDORA-2021-a83da4c566

superm1 commented on contrib/debian/control.in:
```diff
@@ -198,3 +198,9 @@ Depends: ${shlibs:Depends}, ${misc:Depends}, make | build-essential | dpkg-dev
+
+Package: fwupd-qubes-vm-whonix
+Architecture: amd64
+Depends: gpg
```
gpg?  It's not gpg2?

superm1 commented on contrib/debian/fwupd-qubes-vm-whonix.postinst:
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+FWUPD_UPDATEVM_DIR=/home/user/.cache/fwupd
```
`user` is hardcoded to a fixed name?  Doesn't there need to be some sort of lookup process for the correct ID and group?

superm1 replied:
As there is a possibility that someone generates a build with this included and tries to install that `.deb` package, can you please add directory guards to make sure that `/home/user` exists and exit the `postinst` otherwise?

superm1 replied:
Another (better) safety check would be to do a check using:
`getent passwd user`

You can check both the return code as well as the home directory assigned to the user so you can avoid hardcoding anything. I f the return code is non-zero don't run anything else.

Asiderr replied:
I've added safety check.

superm1 commented on contrib/debian/fwupd-qubes-vm-whonix.postrm:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+rm -rf /home/user/.cache/fwupd
```
Likewise here `user` is hardcoded to a fixed name?  Doesn't there need to be some sort of lookup process for the correct ID and group?

superm1 replied:
I think the same safety check I suggested using `getent passwd user` should apply here.

Furthermore, Debian packages aren't supposed to be removing a user's data in the postinst unless called with `purge` as an argument.  So probably you should be checking for that too.

Asiderr replied:
I've added proper checks.

superm1 commented on contrib/fwupd.spec.in:
```diff
@@ -181,6 +183,30 @@ This provides the optional package which is only required on hardware that
+Summary: fwupd wrapper for Qubes OS - dom0 scripts
+Requires:   cabextract
+Requires:   fwupd
+Requires:   gpg
```
Is this the right package name?  Or is it `gnupg2`?

superm1 commented on contrib/qubes/README.md:
```diff
@@ -0,0 +1,243 @@
+
+## OS Requirements
+
+**Operating System:** Qubes OS R4.1 or Qubes OS R4.0
+
+**Admin VM (dom0):** Fedora 25 or higher
+
+**Template VM:** Fedora 30
+
+**Whonix VM:** whonix-gw-15
+
+**fwupd version - dom0:** 0.9.5 or higher
+
+**fwupd version - VMs:** 1.2.6 or higher
+
+qubes-fwupd does not support dom0 updates and downgrades for fwupd 0.9.5 and
+older. Use sys-usb to update external devices.
```
So everything is Fedora based?  If that's the case - let's not worry about any of the Debian packaging stuff.

superm1 replied:
Oh Whonix is Debian based?  Do the changes get pushed up to Debian?

Asiderr replied:
`whonix-gw-15` VM is based on Debian buster

superm1 commented on contrib/qubes/README.md:
````diff
@@ -0,0 +1,243 @@
+5. Install package dependencies
+
+```
+# dnf install cabextract fwudp
````
typo

Asiderr replied:
Fixed

superm1 commented on contrib/qubes/README.md:
````diff
@@ -0,0 +1,243 @@
+9. Install dependencies
+
+```
+# apt install cabextract fwudp
````
typo

Asiderr replied:
Fixed

superm1 commented on contrib/qubes/README.md:
````diff
@@ -0,0 +1,243 @@
+13. Install package dependencies
+
+```
+# qubes-dom0-update cabextract fwudp python36
````
typo

Asiderr replied:
Fixed

superm1 commented on contrib/qubes/doc/heads_udpate.md:
```diff
@@ -0,0 +1,86 @@
+# Heads update
```
file name has a typo in it (`udpate`)

Asiderr replied:
Fixed (changed here and in the file name)

superm1 commented on contrib/qubes/doc/heads_udpate.md:
````diff
@@ -0,0 +1,86 @@
+Currently, you need to pass custom metadata URL:
+
+```
+https://fwupd.org/downloads/firmware-3c81bfdc9db5c8a42c09d38091944bc1a05b27b0.xml.gz
````
is this a leaked private URL from LVFS?   It seems like it.  I don't think we want that in documentation.

Asiderr replied:
Deleted. It won't be needed when the Heads updates will appear in the LVFS production metadata.

superm1 replied:
Perhaps it makes sense on the LVFS side to put them in "testing" channel for now.  And in the documentation you can refer to opting into the testing channel.  When it goes to stable you can change the documentation later on.

superm1 commented on contrib/qubes/doc/uefi_capsule_update.md:
```diff
@@ -0,0 +1,151 @@
+
+### Qubes OS
+
+You need [Qubes R4.1](https://openqa.qubes-os.org/tests/10816/asset/iso/Qubes-4.1-20200726-x86_64.iso)
```
presumably or later?

Asiderr replied:
It was developed and tested on Qubes R4.1. (Link provides the Qubes R4.1 testing iso).

superm1 commented on contrib/qubes/doc/uefi_capsule_update.md:
```diff
@@ -0,0 +1,151 @@
+### Custom dom0 kernel
+
+You have to build a custom linux kernel with
+[patch](../misc/0017-esrt-Add-paravirtualization-support.patch) the issue
```
this seems like a dead URL to me being a relative link

Asiderr replied:
Fixed

superm1 replied:
Still seems like the broken URL to me.

superm1 commented on contrib/debian/control.in:
```diff
@@ -198,3 +198,9 @@ Depends: ${shlibs:Depends}, ${misc:Depends}, make | build-essential | dpkg-dev
  This package is used to control code signing by the Debian signing
  service.
+
+Package: fwupd-qubes-vm-whonix
```
I'm leaning on we should have some sort of flag that determines whether this binary package gets built as part of https://github.com/fwupd/fwupd/blob/master/contrib/ci/generate_debian.py

I don't think it should be done in general purpose Debian or Ubuntu, right?  It should only be when the Debian derivatives Whonix generates the binary packages.

Asiderr replied:
> I don't think it should be done in general purpose Debian or Ubuntu, right?

That's right. It's Qubes specific package. 

> I'm leaning on we should have some sort of flag that determines whether this binary package gets built as part of https://github.com/fwupd/fwupd/blob/master/contrib/ci/generate_debian.py

It could be handled as another OS environment variable e.g. `Debian-x86_64-Whonix` and it would be added to the `TARGET` split: https://github.com/fwupd/fwupd/blob/master/contrib/ci/generate_debian.py#L24

superm1 replied:
Sounds like a good enough approach to me.  If you find while you're doing it that doesn't end up scaling well enough feel free to add `argparse` support there instead though.  The environment parsing was just a convenient way for our first CI jobs to do it.

Asiderr replied:
There is a possibility to generate separate `fedora-qubes` and `debian-x86_64-qubes` containers that build qubes packages. Standard containers don't provide qubes DEBs and RPMs.

superm1 commented on contrib/debian/control.in:
```diff
@@ -198,3 +198,9 @@ Depends: ${shlibs:Depends}, ${misc:Depends}, make | build-essential | dpkg-dev
  service.
+
+Package: fwupd-qubes-vm-whonix
+Architecture: amd64
```
OK.

superm1 commented on contrib/debian/fwupd-qubes-vm-whonix.postinst:
```diff
@@ -0,0 +1,8 @@
+
+mkdir $FWUPD_UPDATEVM_DIR
+chown -R user:user $FWUPD_UPDATEVM_DIR
+chmod +x $FWUPD_UPDATEVM_SCRIPT
```
If it needs to be executable and coming out of `/usr/share` I think it's probably installed to the wrong location.  Shouldn't it be in `/usr/bin/` or `/usr/sbin/` then?

Asiderr replied:
I've moved main executable script (`qubes-fwupdmgr`) to the `/usr/sbin`, also I set read-only permissions for `/usr/share` scripts.

superm1 commented on contrib/debian/fwupd-qubes-vm-whonix.postinst:
```diff
@@ -0,0 +1,8 @@
+FWUPD_UPDATEVM_DIR=/home/user/.cache/fwupd
+FWUPD_UPDATEVM_SCRIPT=/usr/share/qubes-fwupd/fwupd_download_updates.py
+
+mkdir $FWUPD_UPDATEVM_DIR
```
`mkdir -p` would be a lot safer.

Asiderr replied:
Fixed

superm1 commented on contrib/fwupd.spec.in:
```diff
@@ -181,6 +183,30 @@ This provides the optional package which is only required on hardware that
 can be flashed using flashrom. It is probably not required on servers.
 %endif
 
+%package qubes-dom0
```
Similarly to the Debian case I think we want to special case the generation of the `qubes` binary package.  This won't go into real Fedora archives.

hughsie replied:
Completely agree.

Asiderr replied:
There is a possibility to generate separate `fedora-qubes` and `debian-x86_64-qubes` containers that build qubes packages. Standard containers don't provide qubes DEBs and RPMs.

superm1 commented on contrib/qubes/misc/0017-esrt-Add-paravirtualization-support.patch:
```diff
@@ -0,0 +1,65 @@
+Content-Type: text/plain; charset=UTF-8
+Content-Transfer-Encoding: 8bit
+
+Signed-off-by: Norbert Kamiński <norbert.kaminski@3mdeb.com>
```
Before we start carrying this in tree - has it been sent up to the kernel mailing list?  Was it rejected?  Or is it in newer kernels already now?  I think it's better that we refer to an LKML link than a .patch file carried in the fwupd tree.

Asiderr replied:
It wasn't upstreamed, because it is Xen related problem.  [The Xen patch](https://patchwork.kernel.org/project/xen-devel/patch/20200816001949.595424-1-marmarek@invisiblethingslab.com/) was upstreamed but it stuck some time ago. I asked @marmarek if this patch could be added to the [qubes-linux-kernel](https://github.com/QubesOS/qubes-linux-kernel) repo.

superm1 commented on contrib/qubes/test/test_qubes_fwupdmgr.py:
```diff
@@ -0,0 +1,1187 @@
+#!/usr/bin/python3
```
This is missing a license header

Asiderr replied:
Fixed.

superm1 commented on contrib/debian/fwupd-qubes-vm-whonix.install:
```diff
@@ -0,0 +1,2 @@
+usr/share/qubes-fwupd/fwupd_download_updates.py
+usr/share/qubes-fwupd/fwupd_common_vm.py
```
consider moving these to `/usr/bin` or `/bin` instead

Asiderr replied:
I moved those scripts to the `/usr/libexec`. The VM scripts are internals that should be run only by `qubes-fwupdmgr`.

superm1 commented on contrib/debian/control.in:
```diff
@@ -199,3 +199,9 @@ Depends: ${shlibs:Depends}, ${misc:Depends}, make | build-essential | dpkg-dev
  This package is used to control code signing by the Debian signing
  service.
+
+Package: fwupd-qubes-vm-whonix
```
shouldn't this whole section be guarded behind something still?  I would think the package is getting created right now even if qubes is set to false.

Asiderr replied:
@superm1 I've created a separate control.qubes.in for the whonix package. It's included in the `control` if the Qubes environment flag is true.

superm1 commented on contrib/debian/fwupd-qubes-vm-whonix.postinst:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+FWUPD_VM_DIR=/home/user/.cache/fwupd_download_updates
```
better suggestion:
```
getent passwd user | awk '{ split($$0,a,":"); print a[6]}'
```

Asiderr replied:
Implemented.

superm1 commented on contrib/debian/fwupd-qubes-vm-whonix.postrm:
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+if getent passwd user > /dev/null && [ "$1" = "purge" ]; then
```
same suggestion as postinst

Asiderr replied:
Implemented.

hughsie commented on contrib/ci/generate_debian.py:
```diff
@@ -81,7 +86,7 @@ def parse_control_dependencies(requested_type):
                                 suffix = "]"
                             dep = "%s%s%s%s" % (dep, prefix, exclusions[i].text, suffix)
                     deps.append(dep)
-    return deps
+        return deps, QUBES
```
is this right? the indent level changed which made my spider senses tingle

Asiderr replied:
Fixed.

hughsie commented on contrib/fwupd.spec.in:
```diff
@@ -195,6 +198,26 @@ update operation. This subpackage is probably not required on embedded hardware
+%if 0%{?qubes_packages}
+%package qubes-dom0
+Summary: fwupd wrapper for Qubes OS - dom0 scripts
+Requires:   cabextract
```
is this still true? can we use gcab for this instead?

Asiderr replied:
I swapped cabextract to gcab.

hughsie commented on contrib/fwupd.spec.in:
```diff
@@ -195,6 +198,26 @@ update operation. This subpackage is probably not required on embedded hardware
+Summary: fwupd wrapper for Qubes OS - dom0 scripts
+Requires:   cabextract
+Requires:   fwupd
+Requires:   gnupg2
```
same; do we want libcat instead?

Asiderr replied:
Fixed. The current libjcat master must be available as the package in the Debian buster.

hughsie commented on contrib/fwupd.spec.in:
```diff
@@ -501,6 +527,45 @@ done
 
+%if 0%{?qubes_packages}
+%files qubes-vm
+%FWUPD_QUBES_VM_DIR/fwupd_common_vm.py
+%FWUPD_QUBES_VM_DIR/fwupd_download_updates.py
+%FWUPD_QUBES_VM_DIR/fwupd_usbvm_validate.py
```
I'm not sure the `FWUPD_QUBES_VM_DIR` macro adds anything tbh

Asiderr replied:
Replaced with default macros.

hughsie commented on contrib/qubes/README.md:
````diff
@@ -0,0 +1,243 @@
+5. Install package dependencies
+
+```
+# dnf install cabextract fwupd
````
gcab?

Asiderr replied:
Replaced.

hughsie commented on contrib/qubes/README.md:
````diff
@@ -0,0 +1,243 @@
+```
+
+10. Run terminal in the whonix-gw-15 and go to `~/QubesIncoming/qubes-builder`.
+Compare sha sums of the package in TemplateVM and qubes-builder VM. If they
````
SHA

Asiderr replied:
Fixed

hughsie commented on contrib/qubes/README.md:
```diff
@@ -0,0 +1,243 @@
+
+14. Make sure that sys-firewall, sys-whonix, and sys-usb (if exists) are running.
+
+15. Compare the sha sums of the package in dom0 and qubes-builder VM.
```
SHA

Asiderr replied:
Fixed

hughsie commented on contrib/qubes/doc/whonix.md:
````diff
@@ -0,0 +1,30 @@
+$ sudo qubes-fwupdmgr update --whonix
+```
+
+[![asciicast](https://asciinema.org/a/fsqv5Q3Bzi79LoCqoQB2RKts0.svg)](https://asciinema.org/a/fsqv5Q3Bzi79LoCqoQB2RKts0)
````
I do wonder if we want to include these "videos" in the source tree; is asciinema.org going to exist in 5 years?

Asiderr replied:
The oldest video on our blogpost is 2 years old. 

> Is asciinema.org going to exist in 5 years?

That's a good question. I can remove these links if we don't want to carry about them in the future.

superm1 replied:
Maybe a better place to put them is in a blog post or wiki or somewhere easier to spin than a source tree.

hughsie commented on contrib/qubes/meson.build:
```diff
@@ -0,0 +1,55 @@
+             install_dir : 'share/qubes-fwupd/test/logs')
+
+install_data(vm_scripts,
+             install_dir : 'libexec/qubes-fwupd',
+             install_mode : 'rwxrwxr-x')
```
i think the meson standard is two space tabs rather than aligning to the bracket.

Asiderr replied:
I fixed the style in this file.

hughsie commented on contrib/qubes/src/vms/fwupd_download_updates.py:
```diff
@@ -0,0 +1,178 @@
+        self._download_metadata_jcat()
+
+    def _check_shasum(self, file_path, sha):
+        """Compares computed SHA1 checksum with `sha` parameter.
```
I think I have to waive all SHA1 use in RHEL; could we use SHA256 instead?

Asiderr replied:
I replaced SHA1, with SHA256. I can see that some firmware updates don't have SHA256 checksum - e.g. ColorHug2 which I use to test the wrapper. Is it possible to add the SHA256 to this device? Should I add sanity checks if there is SHA256 in the update metadata?

hughsie commented on meson_options.txt:
```diff
@@ -27,6 +27,7 @@ option('plugin_modem_manager', type : 'boolean', value : false, description : 'e
 option('plugin_msr', type : 'boolean', value : true, description : 'enable MSR support')
 option('plugin_flashrom', type : 'boolean', value : false, description : 'enable libflashrom support')
 option('plugin_platform_integrity', type : 'boolean', value : false, description : 'enable platform integrity support')
+option('qubes_enable', type : 'boolean', value : false, description : 'build packages for Qubes OS')
```
I think `qubes` is fine; the _enable_ is implied.

Asiderr replied:
Changed.

superm1 commented on contrib/ci/Dockerfile-debian.in:
```diff
@@ -23,4 +23,4 @@ RUN echo fubar > /etc/machine-id
 %%%INSTALL_DEPENDENCIES_COMMAND%%%
 WORKDIR /github/workspace
-CMD ["./contrib/ci/debian.sh"]
+CMD ["./contrib/ci/debian.sh"%%%QUBES%%%]
```
rather than an argument, can this be done via an environment variable?  That's how we pass `CI` and `CI_NETWORK` and such in.

Asiderr replied:
Deleted. I've added the environment variable `QUBES=true`.

superm1 commented on contrib/ci/Dockerfile-fedora.in:
```diff
@@ -8,4 +8,4 @@ RUN dnf -y update
 %%%INSTALL_DEPENDENCIES_COMMAND%%%
 WORKDIR /github/workspace
-CMD ["./contrib/ci/fedora.sh"]
+CMD ["./contrib/ci/fedora.sh"%%%QUBES%%%]
```
rather than an argument, can this be done via an environment variable? That's how we pass CI and CI_NETWORK and such in.

Asiderr replied:
Deleted. I've added the environment variable `QUBES=true`.

superm1 commented on contrib/ci/fedora.sh:
```diff
@@ -62,7 +69,7 @@ if [ -n "$CI" ]; then
 
 #build RPM packages
-rpmbuild -ba build/fwupd.spec
+rpmbuild -ba --define "qubes_packages $QUBES_MACRO" build/fwupd.spec
```
I would rather see a `qubes_packages` set to zero by default in the spec file instead of "requiring" it set on command line.  Otherwise this spec file can't be used in Fedora anymore by "default".  If you can accomplish this it could become:

```
if [[ $1 == "qubes"* ]]; then
    QUBES="--define \"qubes_packages 1\""
endif
rpmbuild -ba $QUBES build/fwupd.spec
```

Asiderr replied:
Implemented. The change should look a little different:
```
if [ "$QUBES" = "true" ]; then
    QUBES_MACRO=(--define "qubes_packages 1")
fi
...
rpmbuild -ba "${QUBES_MACRO[@]}" build/fwupd.spec
```

superm1 commented on contrib/qubes/test/test_qubes_fwupdmgr.py:
```diff
@@ -0,0 +1,1038 @@
+#
+# Copyright (C) 2021  Norbert Kamiński  <norbert.kaminski@3mdeb.com>
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
```
Would it be possible to license this and everything else LGPL instead?  The rest of the project is so that would be preferable.

Asiderr replied:
I've changed the license to LGPL.

superm1 commented on contrib/qubes/src/qubes_fwupd_heads.py:
```diff
@@ -0,0 +1,150 @@
+HEADS_UPDATES_DIR = os.path.join(BOOT, "updates")
+
+EXIT_CODES = {
+    "ERROR": 1,
```
Do you want to follow the rest of fwupd and also use return code "2" for `NOTHING_TO_DO` perhaps?

Asiderr replied:
Changed.

superm1 commented on contrib/qubes/README.md:
```diff
@@ -0,0 +1,243 @@
+
+## OS Requirements
+
+**Operating System:** Qubes OS R4.1 or Qubes OS R4.0
```
I think you need to update this to 4.1 only since you mentioned elsewhere in this thread you're going to only support that now, and probably the other versions below too then.

Asiderr replied:
Fixed.

superm1 commented on contrib/ci/generate_docker.py:
```diff
@@ -78,6 +80,15 @@ def get_container_cmd():
+            if 'qubes' in TARGET:
+                cmd_line = line.replace(
+                    "%%%QUBES%%%",
+                    ', "-Dqubes_enable=true"'
```
make sure you update this to `-Dqubes=true` when you fix meson options.

Asiderr replied:
Done.

superm1 commented on contrib/ci/generate_docker.py:
```diff
@@ -78,6 +80,15 @@ def get_container_cmd():
+                    ', "-Dqubes_enable=true"'
+                )
+            else:
+                cmd_line = line.replace("%%%QUBES%%%", "")
```
again, preference here is environment variable if possible

Asiderr replied:
Replaced with the environment variable

superm1 commented on contrib/README.md:
```diff
@@ -11,39 +11,66 @@ A Dockerfile for Fedora can be generated in `contrib`.
-docker run --privileged -t -v `pwd`:/build fwupd-debian-x86_64
-docker run --privileged -t -v `pwd`:/build fwupd-debian-i386
-docker run --privileged -t -v `pwd`:/build fwupd-ubuntu-x86_64
+docker run --privileged -t -v `pwd`:/github/workspace fwupd-debian-x86_64
+docker run --privileged -t -v `pwd`:/github/workspace fwupd-debian-i386
+docker run --privileged -t -v `pwd`:/github/workspace fwupd-ubuntu-x86_64
```
were these wrong?  If so, please split this to a separate commit

Asiderr replied:
Done.

superm1 commented on contrib/README.md:
```diff
@@ -11,39 +11,66 @@ A Dockerfile for Fedora can be generated in `contrib`.
-OS=debian-x86_64 ./generate_docker.py
-OS=debian-i386 ./generate_docker.py
-OS=ubuntu-x86_64 ./generate_docker.py
+OS=debian-x86_64 ./generate_docker.py build
+OS=debian-i386 ./generate_docker.py build
+OS=ubuntu-x86_64 ./generate_docker.py build
```
were these missing?  If so, please split this to a separate commit

Asiderr replied:
Done.

Asiderr commented:
@hughsie @superm1 I think it's idea good to make some kind of roadmap dependencies that must be met to merge this pull request with fwupd/fwupd tree:

**Main dependencies**:
- [ ] The libjcat 0.1.6 release must be available in the Debian Buster (DEB) and Fedora 32 (RPM)
- [ ] The ColorHug2 firmware updates must provide SHA256 for testing purposes

**UEFI capsule updates**:
- [ ] The ESRT patch must be merged with Qubes Kernel repo or the changes must be upstreamed to the Xen

**Heads updates**
- [ ] The Qubes R4.1 issue must be resolved

hughsie commented:
> The ColorHug2 firmware updates must provide SHA256 for testing purposes

Which version do you need resigned? I don't want to do them all ideally to avoid churn.

Asiderr commented:
@hughsie Ideally one for update (2.0.7) and one for downgrade (2.0.6 or 2.0.5).

Asiderr commented:
>  Which version do you need resigned? I don't want to do them all ideally to avoid churn.

@hughsie Ideally one for update (2.0.7) and one for downgrade (2.0.6 or 2.0.5).

hughsie commented:
@Asiderr: All three done!

superm1 commented on contrib/ci/generate_debian.py:
```diff
@@ -95,16 +97,25 @@ def update_debian_control(target):
 
-    deps = parse_control_dependencies("build")
+    with open(control_qubes_in, 'r') as rfd:
+        qubes_lines = rfd.readlines()
```
This file, maybe it can be simplified.  Suggestion:

```
if QUBES
    control_qubes_in = ....
    with open(control_qubes_in, 'r') as rfd:
        lines += rfd.readlines()
```

Isn't something like that enough?

Asiderr replied:
Implemented.

hughsie commented:
> 3 for Duplication in regular expression character class

There's also this warning from LGTM; but we're getting there now! Nearly there.

Asiderr commented:
@hughsie @superm1 
>  There's also this warning from LGTM; but we're getting there now! Nearly there.

I fixed LGTM problems.

I've dropped GPG verification and now metadata and firmware updates are validated with jcat verification. I'm still waiting for metadata sync (`fwupdagent` still doesn't provide sha256 for ColorHug2 cabinets).

The wrapper has been successfully tested on Qubes R4.1. For test purposes, I've replaced `jcat-tool` in the following lines with the directories of the `jcat-tool` binary built from the current master:
* [fwupd_common_vm.py](https://github.com/fwupd/fwupd/pull/2710/files#diff-54a9c09216af86031116d0f4926f5c8345c3b3c27296c5ac32059836221a9183R84)
* [fwupd_receive_updates.py](https://github.com/fwupd/fwupd/pull/2710/files#diff-75406b905f82fd50a6d9b1cfe76b42b735a197d178d2cf9c4f1fc95fb2c16981R154)

hughsie commented:
> I'm still waiting for metadata sync

This should have happened yesterday, no? "fwupdmgr refresh --force"

Asiderr commented:
@hughsie I guess so, but fwupdagent still doesn't provide the second checksum:
```
$ fwupdmgr refresh --force
Fetching metadata https://cdn.fwupd.org/downloads/firmware.xml.gz
Downloading…             [***************************************]
Fetching signature https://cdn.fwupd.org/downloads/firmware.xml.gz.asc

Successfully downloaded new metadata: 1 local device supported

$ fwupdagent get-updates
{
  "Devices" : [
    {
      "Name" : "ColorHug2",
      "DeviceId" : "983c3cffc6fd36d32b00b62928d30721eaeb93db",
      "Guid" : [
        "2082b5e0-7a64-478a-b1b2-e3404fab6dad",
        "aa4b4156-9732-55db-9500-bf6388508ee3",
        "101ee86a-7bea-59fb-9f89-6b6297ceed3b",
        "2fa8891f-3ece-53a4-adc4-0dd875685f30",
        "182ab329-1976-5d99-a662-ce618d2ccf88"
      ],
      "Summary" : "An open source display colorimeter",
      "Plugin" : "colorhug",
      "Protocol" : "com.hughski.colorhug",
      "Flags" : [
        "updatable",
        "supported",
        "registered",
        "self-recovery"
      ],
      "Vendor" : "Hughski Ltd.",
      "VendorId" : "USB:0x273F",
      "Version" : "2.0.6",
      "VersionFormat" : "triplet",
      "Icons" : [
        "colorimeter-colorhug"
      ],
      "InstallDuration" : 8,
      "Created" : 1612800403,
      "Releases" : [
        {
          "AppstreamId" : "com.hughski.ColorHug2.firmware",
          "RemoteId" : "lvfs",
          "Summary" : "Firmware for the Hughski ColorHug2 Colorimeter",
          "Description" : "<p>This release fixes prevents the firmware returning an error when the remote SHA1 hash was never sent.</p>",
          "Version" : "2.0.7",
          "Filename" : "658851e6f27c4d87de19cd66b97b610d100efe09",
          "Protocol" : "com.hughski.colorhug",
          "Checksum" : [
            "80bddeb898cda5b87d9837e13a9ace19846053bf"
          ],
          "License" : "GPL-2.0+",
          "Size" : 16384,
          "Uri" : "https://fwupd.org/downloads/0a29848de74d26348bc5a6e24fc9f03778eddf0e-hughski-colorhug2-2.0.7.cab",
          "Homepage" : "http://www.hughski.com/",
          "SourceUrl" : "https://github.com/hughski/colorhug2-firmware",
          "Vendor" : "Hughski Limited",
          "Flags" : [
            "is-upgrade"
          ],
          "InstallDuration" : 8
        }
      ]
    }
  ]
}
```

hughsie commented:
> Fetching signature https://cdn.fwupd.org/downloads/firmware.xml.gz.asc

That looks like a very old fwupd version; possibly one without the SHA256 support?

Asiderr commented:
@hughsie The same result on Fedora32, fwupd 1.5.5 (updates-testing). Metadata contains the SHA256 of the `.cab` archive, but fwupdagent doesn't provide the checksum.

hughsie commented:
Hmm, will investigate. https://bodhi.fedoraproject.org/updates/FEDORA-2021-a83da4c566 now exists for libjcat if that helps.

hughsie commented:
> Hmm, will investigate

Ha, so I think we accidentally fixed this in master a few weeks ago when we added the <artifact> support. I Can tag a new tarball release this week if that helps.

Asiderr commented:
> I Can tag a new tarball release this week if that helps.

This will help a lot. I will run the wrapper tests after the release.

superm1 commented:
>Ha, so I think we accidentally fixed this in master a few weeks ago when we added the artifact support. I Can tag a new tarball release this week if that helps.

It would be really useful to get SBAT stuff locked and merged before that if possible.

hughsie commented:
@Asiderr I've built 1.5.7 for F32 just now: https://bodhi.fedoraproject.org/updates/FEDORA-2021-449618a8a8 -- can you rebase this patchset, do the testing to fix up the remaining issues, and then I think we should aim to get this upstream before it bitrots any more than it has already.

Asiderr commented:
@hughsie Tomorrow I'll test the newest changes and we can focus on the upstream. Also until the end of this week I want to push fwupd for BSD fixes.

hughsie reviewed (approved):
@superm1 can you verify the Debian stuff please.

Asiderr commented:
@hughsie Wrapper is working fine. I added some minor fixes to the tests and docs. I'll push force one more change in a minute. The RPM packages must require specific versions of the libjcat and fwupd.
